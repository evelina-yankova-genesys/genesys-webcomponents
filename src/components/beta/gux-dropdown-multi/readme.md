# gux-dropdown-multi

<!-- Auto Generated Below -->


## Properties

| Property      | Attribute     | Description | Type      | Default     |
| ------------- | ------------- | ----------- | --------- | ----------- |
| `disabled`    | `disabled`    |             | `boolean` | `false`     |
| `filterable`  | `filterable`  |             | `boolean` | `false`     |
| `placeholder` | `placeholder` |             | `string`  | `undefined` |
| `required`    | `required`    |             | `boolean` | `false`     |
| `value`       | `value`       |             | `string`  | `undefined` |


## Methods

### `getSelectedValues() => Promise<string[]>`

Gets the currently selected values.

#### Returns

Type: `Promise<string[]>`




## Slots

| Slot | Description                                                  |
| ---- | ------------------------------------------------------------ |
|      | for a gux-listbox-multi containing gux-option-multi children |


## Dependencies

### Depends on

- [gux-dropdown-multi-tag](gux-dropdown-multi-tag)
- [gux-icon](../../stable/gux-icon)
- [gux-popup](../../stable/gux-popup)

### Graph
```mermaid
graph TD;
  gux-dropdown-multi-beta --> gux-dropdown-multi-tag
  gux-dropdown-multi-beta --> gux-icon
  gux-dropdown-multi-beta --> gux-popup
  gux-dropdown-multi-tag --> gux-icon
  style gux-dropdown-multi-beta fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
