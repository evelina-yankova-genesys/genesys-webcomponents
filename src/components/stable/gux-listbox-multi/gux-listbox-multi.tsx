import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  Host,
  JSX,
  Listen,
  Method,
  Prop,
  State,
  Watch
} from '@stencil/core';

import {
  actOnActiveOption,
  clearActiveOptions,
  goToOption,
  hasPreviousOption,
  hasNextOption,
  onClickedOption,
  setFirstOptionActive,
  setInitialActiveOption,
  setLastOptionActive,
  setNextOptionActive,
  setPreviousOptionActive
} from '../gux-listbox/gux-listbox.service';

import { buildI18nForComponent, GetI18nValue } from '../../../i18n';
import { whenEventIsFrom } from '../../../utils/dom/when-event-is-from';
import simulateNativeEvent from '../../../utils/dom/simulate-native-event';
import { trackComponent } from '../../../usage-tracking';

import translationResources from './i18n/en.json';

/**
 * @slot - collection of gux-option-multi elements
 */
@Component({
  styleUrl: 'gux-listbox-multi.less',
  tag: 'gux-listbox-multi',
  shadow: true
})
export class GuxListboxMulti {
  private i18n: GetI18nValue;

  @Element()
  root: HTMLGuxListboxElement;

  @Prop({ mutable: true })
  value: string;

  @Prop()
  filter: string = '';

  @State()
  selectedValues: string[] = [];

  @State()
  listboxOptions: HTMLGuxOptionElement[] = [];

  @State()
  allListboxOptionsFiltered: boolean;

  @Event()
  internallistboxoptionsupdated: EventEmitter;

  @Listen('focus')
  onFocus(): void {
    setInitialActiveOption(this.root);
  }

  @Listen('blur')
  onBlur(): void {
    clearActiveOptions(this.root);
  }

  @Listen('keydown')
  onKeydown(event: KeyboardEvent): void {
    switch (event.key) {
      case 'Enter':
        event.preventDefault();
        actOnActiveOption(this.root, value => this.updateValue(value));
        return;

      case 'ArrowDown':
        event.preventDefault();
        if (hasNextOption(this.root)) {
          event.stopPropagation();
          setNextOptionActive(this.root);
        } else {
          setFirstOptionActive(this.root);
        }
        return;

      case 'ArrowUp': {
        event.preventDefault();
        if (hasPreviousOption(this.root)) {
          event.stopPropagation();
          setPreviousOptionActive(this.root);
        } else {
          setLastOptionActive(this.root);
        }
        return;
      }

      case 'Home': {
        event.preventDefault();
        setFirstOptionActive(this.root);
        return;
      }

      case 'End': {
        event.preventDefault();
        setLastOptionActive(this.root);
        return;
      }

      case ' ': {
        event.preventDefault();
        return;
      }
    }

    if (event.key.length === 1) {
      goToOption(this.root, event.key);
      return;
    }
  }

  @Listen('keyup')
  onKeyup(event: KeyboardEvent): void {
    switch (event.key) {
      case ' ':
        actOnActiveOption(this.root, value => this.updateValue(value));
        return;
    }
  }

  @Listen('mousemove', {})
  onMousemove(): void {
    clearActiveOptions(this.root);
  }

  @Listen('click')
  onClick(event: MouseEvent): void {
    whenEventIsFrom(
      'gux-option-multi',
      event,
      (option: HTMLGuxOptionMultiElement) => {
        onClickedOption(option, value => this.updateValue(value));
      }
    );
  }

  @Watch('value')
  validateValue() {
    if (this.value) {
      this.selectedValues = this.value.split(',');
    } else {
      this.selectedValues = [];
    }
  }

  // eslint-disable-next-line @typescript-eslint/require-await
  @Method()
  async guxSelectActive(): Promise<void> {
    actOnActiveOption(this.root, value => this.updateValue(value));
  }

  private setListboxOptions(): void {
    if (this.value) {
      this.selectedValues = this.value.split(',');
    }
    this.listboxOptions = Array.from(
      this.root.children
    ) as HTMLGuxOptionElement[];
    this.internallistboxoptionsupdated.emit();
  }

  private updateValue(newValue: string): void {
    if (!this.selectedValues.includes(newValue)) {
      const newArray = [...this.selectedValues, newValue];
      this.selectedValues = newArray;
      this.value = this.selectedValues.join(',');
    } else {
      this.selectedValues = this.selectedValues.filter(e => e !== newValue);
      this.value = this.selectedValues.join(',');
    }
    simulateNativeEvent(this.root, 'input');
    simulateNativeEvent(this.root, 'change');
  }

  async componentWillLoad(): Promise<void> {
    trackComponent(this.root);
    this.i18n = await buildI18nForComponent(this.root, translationResources);

    this.setListboxOptions();
  }

  componentWillRender(): void {
    this.listboxOptions.forEach(listboxOption => {
      listboxOption.selected = this.selectedValues.includes(
        listboxOption.value
      );
      listboxOption.filtered = !listboxOption.textContent
        .toLowerCase()
        .startsWith(this.filter.toLowerCase());
    });

    this.allListboxOptionsFiltered =
      this.listboxOptions.filter(listboxOption => !listboxOption.filtered)
        .length === 0;
  }

  renderAllListboxOptionsFiltered(): JSX.Element {
    if (this.allListboxOptionsFiltered) {
      return (
        <div class="gux-no-matches">{this.i18n('noMatches')}</div>
      ) as JSX.Element;
    }
  }

  render(): JSX.Element {
    return (
      <Host role="listbox" aria-multiselectable="true" tabindex={0}>
        <slot onSlotchange={() => this.setListboxOptions()} />
        {this.renderAllListboxOptionsFiltered()}
      </Host>
    ) as JSX.Element;
  }
}
