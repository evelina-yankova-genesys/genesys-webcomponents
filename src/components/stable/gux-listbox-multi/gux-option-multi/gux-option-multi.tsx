import { Component, Element, h, Host, JSX, Listen, Prop } from '@stencil/core';

import { randomHTMLId } from '../../../../utils/dom/random-html-id';

@Component({
  styleUrl: 'gux-option-multi.less',
  tag: 'gux-option-multi'
})
export class GuxOptionMulti {
  @Element()
  root: HTMLElement;

  @Prop()
  value: string;

  @Prop()
  active: boolean = false;

  @Prop()
  selected: boolean = false;

  @Prop()
  disabled: boolean = false;

  @Prop()
  filtered: boolean = false;

  @Prop({ mutable: true })
  hovered: boolean = false;

  @Listen('mouseenter')
  onmouseenter() {
    this.hovered = true;
  }

  @Listen('mouseleave')
  onMouseleave() {
    this.hovered = false;
  }

  componentWillLoad(): void {
    this.root.id = this.root.id || randomHTMLId('gux-option-multi');
  }

  private getAriaSelected(): boolean | string {
    if (this.disabled) {
      return false;
    }

    return this.selected ? 'true' : 'false';
  }

  render(): JSX.Element {
    return (
      <Host
        role="option"
        class={{
          'gux-active': this.active,
          'gux-disabled': this.disabled,
          'gux-filtered': this.filtered,
          'gux-hovered': this.hovered,
          'gux-selected': this.selected
        }}
        aria-selected={this.getAriaSelected()}
        aria-disabled={this.disabled.toString()}
      >
        <div class="gux-option">
          <slot />
        </div>
      </Host>
    ) as JSX.Element;
  }
}
