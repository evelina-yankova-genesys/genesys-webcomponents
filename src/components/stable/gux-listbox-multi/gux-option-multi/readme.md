# gux-option-multi

<!-- Auto Generated Below -->


## Properties

| Property   | Attribute  | Description | Type      | Default     |
| ---------- | ---------- | ----------- | --------- | ----------- |
| `active`   | `active`   |             | `boolean` | `false`     |
| `disabled` | `disabled` |             | `boolean` | `false`     |
| `filtered` | `filtered` |             | `boolean` | `false`     |
| `hovered`  | `hovered`  |             | `boolean` | `false`     |
| `selected` | `selected` |             | `boolean` | `false`     |
| `value`    | `value`    |             | `string`  | `undefined` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
