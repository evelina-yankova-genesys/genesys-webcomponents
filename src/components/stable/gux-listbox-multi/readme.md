# gux-listbox-multi

<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description | Type     | Default     |
| -------- | --------- | ----------- | -------- | ----------- |
| `filter` | `filter`  |             | `string` | `''`        |
| `value`  | `value`   |             | `string` | `undefined` |


## Events

| Event                           | Description | Type               |
| ------------------------------- | ----------- | ------------------ |
| `internallistboxoptionsupdated` |             | `CustomEvent<any>` |


## Methods

### `guxSelectActive() => Promise<void>`



#### Returns

Type: `Promise<void>`




## Slots

| Slot | Description                             |
| ---- | --------------------------------------- |
|      | collection of gux-option-multi elements |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
