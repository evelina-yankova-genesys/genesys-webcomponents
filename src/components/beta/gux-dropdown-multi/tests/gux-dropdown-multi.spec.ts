import { newSpecPage } from '@stencil/core/testing';
import MutationObserver from 'mutation-observer';

import { GuxDropdownMulti } from '../gux-dropdown-multi';
import { GuxListboxMulti } from '../../../stable/gux-listbox-multi/gux-listbox-multi';
import { GuxOptionMulti } from '../../../stable/gux-listbox-multi/gux-option-multi/gux-option-multi';

const components = [GuxDropdownMulti, GuxListboxMulti, GuxOptionMulti];
const html = `
<gux-dropdown-multi-beta>
  <gux-listbox-multi aria-label="Animals">
    <gux-option-multi value="a" disabled>Ant</gux-option-multi>
    <gux-option-multi value="b">Bat</gux-option-multi>
    <gux-option-multi value="c">Cat</gux-option-multi>
  </gux-listbox-multi>
</gux-dropdown-multi-beta>
`;
const language = 'en';

describe('gux-dropdown', () => {
  beforeEach(() => {
    (
      global as NodeJS.Global & {
        MutationObserver: any;
      }
    ).MutationObserver = MutationObserver;
  });

  describe('#render', () => {
    it(`should render as expected`, async () => {
      const page = await newSpecPage({ components, html, language });

      expect(page.rootInstance).toBeInstanceOf(GuxDropdownMulti);
      expect(page.root).toMatchSnapshot();
    });
  });
});
