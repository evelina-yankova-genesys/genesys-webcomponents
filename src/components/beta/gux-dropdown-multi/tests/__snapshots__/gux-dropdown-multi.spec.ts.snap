// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-dropdown #render should render as expected 1`] = `
<gux-dropdown-multi-beta>
  <mock:shadow-root>
    <div class="gux-dropdown-container">
      <gux-popup>
        <div class="gux-target-container gux-target-container-collapsed" slot="target">
          <button aria-expanded="false" aria-haspopup="listbox" class="gux-field gux-field-button" type="button">
            <div class="gux-field-content">
              <div class="gux-placeholder">
                Select...
              </div>
            </div>
            <gux-icon class="gux-expand-icon" iconname="chevron-small-down" screenreader-text="Dropdown"></gux-icon>
          </button>
        </div>
        <div class="gux-listbox-container" slot="popup">
          <slot></slot>
        </div>
      </gux-popup>
    </div>
  </mock:shadow-root>
  <gux-listbox-multi aria-label="Animals" aria-multiselectable="true" role="listbox" tabindex="0">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    <gux-option-multi aria-disabled="true" class="gux-disabled" disabled="" id="gux-option-multi-i" role="option" value="a">
      <!---->
      <div class="gux-option">
        Ant
      </div>
    </gux-option-multi>
    <gux-option-multi aria-disabled="false" aria-selected="false" id="gux-option-multi-i" role="option" value="b">
      <!---->
      <div class="gux-option">
        Bat
      </div>
    </gux-option-multi>
    <gux-option-multi aria-disabled="false" aria-selected="false" id="gux-option-multi-i" role="option" value="c">
      <!---->
      <div class="gux-option">
        Cat
      </div>
    </gux-option-multi>
  </gux-listbox-multi>
</gux-dropdown-multi-beta>
`;
